module/retrieval/bitswap.go

```diff
@@ -0,0 +1,248 @@
+
+type BitswapRetriever struct {
+	log          zerolog.Logger
+	done         chan interface{}
+	bitswap      func() gocar.ReadStore
+	events       []TimeEventPair
+	cidDurations map[cid.Cid]time.Duration
+	traverser    gocar.SelectiveCarPrepared
+	startTime    time.Time
+}
+
+type BitswapRetrieverBuilder struct{}
+
+type bitswapAdapter struct {
+	session *bswap.Session
+}
+
+func (bi bitswapAdapter) Get(ctx context.Context, c cid.Cid) (blocks.Block, error) {
+	bytes, err := bi.session.Get(c)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot get bytes from adapter")
+	}
+
+	block, err := blocks.NewBlockWithCid(bytes, c)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot create block from adapter")
+	}
+
+	return block, nil
+}
+
+// github.com/willscott/go-selfish-bitswap-client has a hardcoded timeout for Session
+// of 10 seconds. Because of this, we need to make a new session for each block if we want
+// the Dump process to take longer than 10 seconds in total. This is accomplished by using
+// the bitswapCallback function to create a new session for each block.
+func (b *BitswapRetrieverBuilder) Build(
+	ctx context.Context,
+	minerInfo *module.MinerInfoResult,
+) (*BitswapRetriever, Cleanup, error) {
+	libp2p, err := role.NewLibp2pHostWithRandomIdentityAndPort()
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "cannot create libp2p host")
+	}
+
+	bitswapCallback := func() gocar.ReadStore {
+		return bitswapAdapter{
+			session: bswap.New(libp2p, *minerInfo.PeerID),
+		}
+	}
+
+	// nolint:exhaustruct
+	return &BitswapRetriever{
+			log:          log.With().Str("role", "retrieval_bitswap").Caller().Logger(),
+			bitswap:      bitswapCallback,
+			done:         make(chan interface{}),
+			events:       make([]TimeEventPair, 0),
+			cidDurations: make(map[cid.Cid]time.Duration),
+			startTime:    time.Time{},
+		}, func() {
+			libp2p.Close()
+		}, nil
+}
+
+func (b *BitswapRetriever) Type() task.Type {
+	return task.Retrieval
+}
+
+// callback gets called after a successful block retrieval during traverser.Dump.
+func (b *BitswapRetriever) onNewCarBlock(block gocar.Block) error {
+	// nolint:exhaustruct
+	event := TimeEventPair{
+		Timestamp: time.Now(),
+		Received:  block.Size,
+	}
+
+	if len(b.events) == 0 {
+		event.Message = fmt.Sprintf("first-bytes received: [Block %s]", block.BlockCID)
+		event.Code = string(FirstByteReceived)
+	} else {
+		event.Message = fmt.Sprintf("bytes received: [Block %s]", block.BlockCID)
+		event.Code = string(BlockReceived)
+	}
+
+	b.events = append(b.events, event)
+
+	return nil
+}
+
+// nolint:lll
+func (b *BitswapRetriever) Retrieve(ctx context.Context, root cid.Cid, timeout time.Duration) (
+	*ResultContent,
+	error,
+) {
+	dag := gocar.Dag{Root: root, Selector: selectorparse.CommonSelector_ExploreAllRecursively}
+
+	car := gocar.NewSelectiveCar(ctx, b, []gocar.Dag{dag}, gocar.TraverseLinksOnlyOnce())
+
+	traverser, err := car.Prepare(b.onNewCarBlock)
```

Comment:
the reason I was not dependending on `car.Perpare` for the traversal that it does is because I did not think it allowed for a timeout context, however, upon further digging, I found that the `car.ctx` is passed into the `gocar.traverse` function through the `ipld.LinkContext` [here](https://github.com/ipld/go-car/blob/e6f244225a16bab6a5d1d8adcb514660f644ed9b/selectivecar.go#L265)

this means that Prepare could be the actual traversal that the blocks are fetched on, I am wondering if I should change it to rely on Prepare vs Dump?  

Dump is probably using some type of caching from the previously fetched blocks.

Reply:
Oh, just go a step lower. I was suggesting you copy how `NewSelectiveCar` is implemented **internally**. Don't use the actual car library here itself. Once you crack open the hood of go-car the answer to your initial question will be obvious.